nixos/modules/services/networking/firewall.nix

```diff
@@ -510,14 +512,43 @@ in
 
     networking.firewall.extraAllowed = mkOption {
       type = types.listOf (types.submodule { options = {
-           source = mkOption { type = types.string; default = ""; example = "example.com"; description = "Source of the packets"; };
-           protocol = mkOption { type = types.enum ["tcp" "udp" "icmp"]; default = ""; example = "tcp"; description = "Protocol used."; };
-           dport = mkOption { type = types.int; default = 0; example = 22; description = "Destination port of the packet."; };
-           useIPv4 = mkOption { type = types.bool; default = true; description = "Use the IPv4 address of this host."; };
-           useIPv6 = mkOption { type = types.bool; default = true; description = "Use the IPv6 address of this host."; };
+           sourceAddress = mkOption {
+                type = types.nullOr types.string;
+                default = null;
+                example = "198.51.100.0";
+                description = "Source of the packets. According to the iptables manual, DNS host names might be used here, but their use is discouraged (in particular because this delegates the security rule to DNS servers). DNS host names are resolved only when the rule is loaded.";
+           };
+           destAddress = mkOption {
+                type = types.nullOr types.string;
+                default = null;
+                example = "198.51.100.0";
+                description = "Destination of the packets. Same comment as for the `sourceAddress` parameter.";
+           };
+           protocols = mkOption {
+                type = types.nullOr (types.listOf (types.enum ["tcp" "udp" "icmp"]));
+                default = null;
+                example = ["tcp"];
+                description = "Protocol used.";
+           };
+           sport = mkOption {
+                type = types.nullOr types.string;
+                default = null;
+                example = null;
+                description = "Source port of the packet. This might be either an integer port number, or the name of a protocol, such as \"http\" or \"ssh\".";
+           };
+           dport = mkOption {
+                type = types.nullOr types.string;
+                default = null;
+                example = "22";
+                description = "Destination port of the packet. This might be either an integer port number, or the name of a protocol, such as \"http\" or \"ssh\".";
+           };
+           addressFamily = mkOption {
+                type = types.listOf (types.enum ["IPv4" "IPv6"]);
```

Comment:
All instances of `types.enum` I found, use  lowercase. So I think it would be consistent to keep that

```
type = types.listOf (types.enum ["ipv4" "ipv6"]);
```